bgpd: add support for SRv6 L3VPN for IPv4 and IPv6 address families using a single SID

cscarpitta opened the pull request:
This PR introduces the following changes in FRR:

* add new optional CLI arguments `block-len` and `node-len` to make the block length and node length of an SRv6 locator configurable
* add support for SRv6 End.DT46 behavior (both in _zebra_ and _sharpd_)
* add a CLI command to configure a per-VRF SRv6 SID in bgpd (used to advertise a L3VPN for IPv4 and IPv6 address families  using a single SID (End.DT46)
* add block/node/func/arg length to the output returned by `show bgp segment-routing srv6`
* add per-VRF SRv6 SID to the output returned by `show bgp segment-routing srv6`
* change bgpd to advertise block/node/func/arg length configured in the SRv6 locator instead of the hardcoded default values
* add a new topotest for SRv6 L3VPN for IPv4 and IPv6 address families using a single SID
* add a new topotest to verify the `block-len` and `node-len` parameters of an SRv6 locator
* add a new topotest to verify the End.DT46 seg6local routes
* add documentation for all new features (bgpd, sharpd, and zebra)

Signed-off-by: Carmine Scarpitta carmine.scarpitta@uniroma2.it

cscarpitta commented:
This PR includes some features already proposed in PR #11348. We waited a long time for the PR #11348 to be merged, but we saw that there has been no progress for some time. If the PR #11348 is merged in the meantime we are ready to rebase and resubmit our changes.

donaldsharp commented:
The other PR is not being merged because you haven't answered @ton31337 's question yet.  You need to work it out with him.  All in all I agree with him, I don't understand what that function is trying to do or why.

cscarpitta commented:
Hi @donaldsharp ,

thanks for your comment.

Let me give some context.

I'm not the author of the other PR. I got aware of it when I was preparing this PR. Anyway, some of the features included in the other PR are already supported in my PR.

My PR is part of the work on the integration of FRR and SONiC to support SRv6 ( https://github.com/sonic-net/SONiC/blob/master/doc/srv6/srv6_hld.md). 

The SRv6 Network Programming model allows the support of L3VPN services for IPv4 and IPv6 address families using a single SRv6 SID (https://datatracker.ietf.org/doc/html/rfc8986#section-4.8).

In the current implementation of FRR/bgpd, you can export VPN routes using a per-address-family SRv6 SID (a SID for the IPv4 routes, a second SID for the IPv6 routes).

In my PR, I extended bgpd to export VPN routes for both IPv4 and IPv6 address families using a single SID (instead of using two SIDs). This is in alignment with SRv6 Network Programming model (RFC8986). 

As part of my PR, I added some updates to zebra and bgpd including: 
* Update bgpd to advertise block/node/func/arg length configured in the SRv6 locator instead of the hardcoded default values (https://datatracker.ietf.org/doc/html/rfc9252#section-3.2.1). 

This is the only change that overlaps between my PR and the other PR (#11348). All the other changes proposed in my PR are not included in the other PR. 

Regards,
Carmine

ton31337 commented on bgpd/bgp_vty.c:
```diff
@@ -17700,6 +17811,13 @@ int bgp_config_write(struct vty *vty)
 
+		tovpn_sid_index = bgp->tovpn_sid_index;
+		if (CHECK_FLAG(bgp->vrf_flags, BGP_VRF_TOVPN_SID_AUTO)) {
+			vty_out(vty, " sid vpn per-vrf export %s\n", "auto");
```
Just `auto` without string interpolation? :)

cscarpitta replied:
Exactly :) Fixed in https://github.com/FRRouting/frr/pull/11673/commits/b1e7ed5181cef3e08a6dc72e07ca3cc02969292c.

ton31337 commented on bgpd/bgp_zebra.c:
```diff
@@ -3260,6 +3260,17 @@ static int bgp_zebra_process_srv6_locator_delete(ZAPI_CALLBACK_ARGS)
+		tovpn_sid = bgp_vrf->tovpn_sid;
+		if (tovpn_sid) {
+			tmp_prefi.family = AF_INET6;
+			tmp_prefi.prefixlen = 128;
```
IPV6_MAX_BITLEN

cscarpitta replied:
Fixed in https://github.com/FRRouting/frr/pull/11673/commits/059820c65586b51705aabb11b47d59dfa379ea96.

ton31337 reviewed (changes requested):
Quick review from me + CI warnings should be fixed too.

cscarpitta commented on bgpd/bgp_mplsvpn.c:
```diff
@@ -405,11 +405,81 @@ void vpn_leak_zebra_vrf_sid_update(struct bgp *bgp, afi_t afi)
+	}
+
+	if (debug) {
+		inet_ntop(AF_INET6, tovpn_sid, buf, sizeof(buf));
```
Fixed in https://github.com/FRRouting/frr/pull/11673/commits/7150170214d280047a83c5b9b06b24ee67e56a02.

cscarpitta commented on bgpd/bgp_mplsvpn.c:
```diff
@@ -635,6 +745,83 @@ void ensure_vrf_tovpn_sid(struct bgp *bgp_vpn, struct bgp *bgp_vrf, afi_t afi)
+	}
+
+	if (debug) {
+		inet_ntop(AF_INET6, tovpn_sid, buf, sizeof(buf));
```
Fixed in https://github.com/FRRouting/frr/pull/11673/commits/7150170214d280047a83c5b9b06b24ee67e56a02.

cscarpitta commented on bgpd/bgp_mplsvpn.c:
```diff
@@ -635,6 +745,83 @@ void ensure_vrf_tovpn_sid(struct bgp *bgp_vpn, struct bgp *bgp_vrf, afi_t afi)
+						  tovpn_sid_locator, tovpn_sid);
+
+	if (tovpn_sid_transpose_label == 0) {
+		zlog_debug("%s: not allocated new sid for vrf %s", __func__,
```
Fixed in https://github.com/FRRouting/frr/pull/11673/commits/a451cd5bec031263db50c32d0ac9ce2651cc005c.

cscarpitta commented on bgpd/bgp_vty.c:
```diff
@@ -8833,6 +8846,90 @@ DEFPY (af_sid_vpn_export,
+	int idx = 0;
+	bool yes = true;
+
+	if (argv_find(argv, argc, "no", &idx))
```
Fixed in https://github.com/FRRouting/frr/pull/11673/commits/881b3a5bdf50081f5e6eed283f0f4635684887b0. Thanks for the suggestion.

cscarpitta commented on bgpd/bgp_mplsvpn.c:
```diff
@@ -635,6 +745,83 @@ void ensure_vrf_tovpn_sid(struct bgp *bgp_vpn, struct bgp *bgp_vrf, afi_t afi)
+
+	if (debug) {
+		inet_ntop(AF_INET6, tovpn_sid, buf, sizeof(buf));
+		zlog_debug("%s: new sid %s allocated for vrf %s", __func__, buf,
```
Here the debug guard is present at line 800, isn't it?

cscarpitta commented:
Hi @ton31337, many thanks for your review.

I addressed all your comments and fixed most of the warnings reported by frrbot and CI.

Regarding the CI, 3 style warnings remain unresolved. These warnings are related to the `rt_netlink.c` file. To fix them, we need a code refactor that goes beyond this PR, because the warnings are related to code that was already present. In my opinion, it is more appropriate to open a separate pull request to fix these style issues.

Concerning the warnings reported by frrbot, some of these are related to `bgp_vty.c` and `zebra_vty.c`, and I think these warnings are going in the wrong direction. The others are related to header files taken from the Linux kernel and (in my opinion) it is not worth fixing them.

Let me know what you think.

ton31337 commented:
@cscarpitta could you squash or amend the last commits to the right ones because they are just garbage (a burden to backport/revert things in the future)?

cscarpitta commented:
> @cscarpitta could you squash or amend the last commits to the right ones because they are just garbage (a burden to backport/revert things in the future)?

Done! I squashed both the last and style-related commits to the right ones and force-pushed the branch.

donaldsharp commented on doc/user/bgp.rst:
```diff
@@ -2841,6 +2841,23 @@ L3VPN SRv6
+Configuration of the SRv6 SID used to advertise a L3VPN for both IPv4 and IPv6
+is accomplished via the following command in the context of a VRF:
+
+.. clicmd:: sid vpn per-vrf export (1..255)|auto
```
is this truly only for export?

cscarpitta replied:
Yes, it is only for export.

``sid vpn per-vrf export (1..255)|auto`` only enables exporting the routes from a VRF to the VPN RIB using a given SRv6 SID. This command has no impact on the imported routes. In order to enable and configure the import of routes from the VPN RIB to a VRF, you need to use the ``import vpn`` and ``rt vpn import`` commands already supported in bgpd (in the context of an address-family). 
Note: this command already exists before. It was used to do export per-address family. We reuse the same syntax to do export per-vrf.

donaldsharp commented on doc/user/zebra.rst:
```diff
@@ -757,10 +757,20 @@ and this section also helps that case.
    function bits range, they can configure with ``func-bits``
    option.
 
+   The ``block-len`` parameter allows the user to configure the length of the
```
might be worth explaining how the values must relate to each other?

cscarpitta replied:
Thanks for the suggestion. I added this to the zebra documentation. Please find a copy of the documentation below.

> .. clicmd:: prefix X:X::X:X/M [func-bits 32] [block-len 40] [node-len 24]
> 
>    Set the ipv6 prefix block of the locator. SRv6 locator is defined by
>    RFC8986. The actual routing protocol specifies the locator and allocates a
>    SID to be used by each routing protocol. This SID is included in the locator
>    as an IPv6 prefix.
> 
>    Following example console log shows the typical configuration of SRv6
>    data-plane. After a new SRv6 locator, named loc1, is created, loc1's prefix
>    is configured as ``2001:db8:1:1::/64``.  If user or some routing daemon
>    allocates new SID on this locator, new SID will allocated in range of this
>    prefix. For example, if some routing daemon creates new SID on locator
>    (``2001:db8:1:1::/64``), Then new SID will be ``2001:db8:1:1:7::/80``,
>    ``2001:db8:1:1:8::/80``, and so on.  Each locator has default SID that is
>    SRv6 local function "End".  Usually default SID is allocated as
>    ``PREFIX:1::``.  (``PREFIX`` is locator's prefix) For example, if user
>    configure the locator's prefix as ``2001:db8:1:1::/64``, then default SID
>    will be ``2001:db8:1:1:1::``)
> 
>    This command takes three optional parameters: ``func-bits``, ``block-len``
>    and ``node-len``. These parameters allow users to set the format for the SIDs
>    allocated from the SRv6 Locator. SID Format is defined in RFC 8986.
> 
>    According to RFC 8986, an SRv6 SID consists of BLOCK:NODE:FUNCTION:ARGUMENT,
>    where BLOCK is the SRv6 SID block (i.e., the IPv6 prefix allocated for SRv6
>    SIDs by the operator), NODE is the identifier of the parent node instantiating
>    the SID, FUNCTION identifies the local behavior associated to the SID and
>    ARGUMENT encodes additional information used to process the behavior.
>    BLOCK and NODE make up the SRv6 Locator.
> 
>    The function bits range is 16bits by default.  If operator want to change
>    function bits range, they can configure with ``func-bits``
>    option.
> 
>    The ``block-len`` and ``node-len`` parameters allow the user to configure the
>    length of the SRv6 SID block and SRv6 SID node, respectively. Both the lengths
>    are expressed in bits.
> 
>    ``block-len``, ``node-len`` and ``func-bits`` may be any value as long as
>    ``block-len+node-len = locator-len`` and ``block-len+node-len+func-bits <= 128``.
> 
>    When both ``block-len`` and ``node-len`` are omitted, the following default
>    values are used: ``block-len = 24``, ``node-len = prefix-len-24``.
> 
>    If only one parameter is omitted, the other parameter is derived from the first.
>

donaldsharp commented on bgpd/bgp_vty.c:
```diff
@@ -8833,6 +8846,86 @@ DEFPY (af_sid_vpn_export,
+		 BGP_DEBUG(vpn, VPN_LEAK_FROM_VRF));
+
+	if (no) {
+		/* implement me */
```
We need the ability to handle the no form of this command

cscarpitta replied:
Done. I updated the PR with the implementation of the no form of the command (commit https://github.com/FRRouting/frr/pull/11673/commits/a5e6a474bb39d8b2201c28b9d9c1d16e483d7079). Also, I extended the topotest to test it (commit https://github.com/FRRouting/frr/pull/11673/commits/81e59c62750343a19e75838fef3913e21056a1b9).

donaldsharp commented on zebra/rt_netlink.c:
```diff
@@ -2713,6 +2723,18 @@ ssize_t netlink_nexthop_msg_encode(uint16_t cmd,
 							    ctx->table))
 							return 0;
 						break;
+					case SEG6_LOCAL_ACTION_END_DT46:
```
I see no code added to decode these types of nexthops from the kernel?

cscarpitta replied:
The code to decode `SEG6_LOCAL_ACTION_END_DT46` nexthops is already present in ``parse_encap_seg6local()``. The `SEG6_LOCAL_ACTION_END_DT46` uses the Netlink attribute `SEG6_LOCAL_VRFTABLE` which is already supported for the `SEG6_LOCAL_ACTION_END_DT4` nexthop.

Pdoijode commented on zebra/zebra_srv6_vty.c:
```diff
@@ -271,35 +275,42 @@ DEFUN (no_srv6_locator,
-	 */
-	locator->block_bits_length = prefix->prefixlen - 24;
-	locator->node_bits_length = 24;
+	/* Resolve optional arguments */
```
Should we do some validation here to ensure that "block_len + node_len + function_len + arg_len" is not > 128 bits? 

Consider the following scenario:

"locator r2
    prefix cafe:3:3::1/128 func-bits 16
   exit"

In case someone configures the above locator (which is a wrong config), then the SID parameters will look something like the following: 
 
  block_bits_length = 104 'h',
  node_bits_length = 24 '\030',
  function_bits_length = 16 '\020',
  argument_bits_length = 0 '\000',
===> Total length is 144!

Here block_len + node_len + function_len + arg_len exceeds 128 bits and BGP will end up sending an update with L3 srv6 sub-sub tlv containing transpose offset = 128(block_len + node_len) and transpose_len = 16 which is wrong.

according to section 3.1 of RFC 8986:
 "This document defines an SRv6 SID as consisting of LOC:FUNCT:ARG,
   where a locator (LOC) is encoded in the L most significant bits of
   the SID, followed by F bits of function (FUNCT) and A bits of
   arguments (ARG).  L, the locator length, is flexible, and an operator
   is free to use the locator length of their choice.  F and A may be
   any value as long as L+F+A <= 128"

cscarpitta replied:
I agree. I updated the PR to add a check to validate the locator (commit https://github.com/FRRouting/frr/pull/11673/commits/983d79618bbd2f8c247b1cd2cbfd3b8b6d6a2fce). We accept a locator only if ``block_len + node_len + function_len + arg_len <= 128``.

donaldsharp commented:
I'm seeing memory leaks when I run the new tests with --valgrind-memleaks:

==63389== 704 (344 direct, 360 indirect) bytes in 1 blocks are definitely lost in loss record 26 of 30
==63389==    at 0x48455EF: calloc (vg_replace_malloc.c:1328)
==63389==    by 0x4910261: qcalloc (memory.c:116)
==63389==    by 0x491208E: srv6_locator_alloc (srv6.c:132)
==63389==    by 0x1EEF05: srv6_locator (zebra_srv6_vty.c:247)
==63389==    by 0x48C9423: cmd_execute_command_real (command.c:997)
==63389==    by 0x48C96FC: cmd_execute_command_strict (command.c:1108)
==63389==    by 0x48C9B94: command_config_read_one_line (command.c:1268)
==63389==    by 0x48C9D09: config_from_file (command.c:1313)
==63389==    by 0x496B831: vty_read_file (vty.c:2347)
==63389==    by 0x496C01F: vty_read_config (vty.c:2567)
==63389==    by 0x48FED1C: frr_config_read_in (libfrr.c:984)
==63389==    by 0x496118F: thread_call (thread.c:2005)
==63389==    by 0x48FF5E4: frr_run (libfrr.c:1198)
==63389==    by 0x1A11F3: main (main.c:475)
==63389==
==63389== 704 (344 direct, 360 indirect) bytes in 1 blocks are definitely lost in loss record 27 of 30
==63389==    at 0x48455EF: calloc (vg_replace_malloc.c:1328)
==63389==    by 0x4910261: qcalloc (memory.c:116)
==63389==    by 0x491208E: srv6_locator_alloc (srv6.c:132)
==63389==    by 0x1EEF05: srv6_locator (zebra_srv6_vty.c:247)
==63389==    by 0x48C9423: cmd_execute_command_real (command.c:997)
==63389==    by 0x48C9BBC: command_config_read_one_line (command.c:1280)
==63389==    by 0x48C9D09: config_from_file (command.c:1313)
==63389==    by 0x496B831: vty_read_file (vty.c:2347)
==63389==    by 0x496C01F: vty_read_config (vty.c:2567)
==63389==    by 0x48FED1C: frr_config_read_in (libfrr.c:984)
==63389==    by 0x496118F: thread_call (thread.c:2005)
==63389==    by 0x48FF5E4: frr_run (libfrr.c:1198)
==63389==    by 0x1A11F3: main (main.c:475)
==63389==
==63389== 704 (344 direct, 360 indirect) bytes in 1 blocks are definitely lost in loss record 28 of 30
==63389==    at 0x48455EF: calloc (vg_replace_malloc.c:1328)
==63389==    by 0x4910261: qcalloc (memory.c:116)
==63389==    by 0x491208E: srv6_locator_alloc (srv6.c:132)
==63389==    by 0x1EEF05: srv6_locator (zebra_srv6_vty.c:247)
==63389==    by 0x48C9423: cmd_execute_command_real (command.c:997)
==63389==    by 0x48C9590: cmd_execute_command (command.c:1057)
==63389==    by 0x48C9AEA: cmd_execute (command.c:1224)
==63389==    by 0x49676F3: vty_command (vty.c:482)
==63389==    by 0x496948C: vty_execute (vty.c:1245)
==63389==    by 0x496B2B4: vtysh_read (vty.c:2144)
==63389==    by 0x496118F: thread_call (thread.c:2005)
==63389==    by 0x48FF5E4: frr_run (libfrr.c:1198)
==63389==    by 0x1A11F3: main (main.c:475)
==63426== 56 bytes in 1 blocks are definitely lost in loss record 11 of 38
==63426==    at 0x48455EF: calloc (vg_replace_malloc.c:1328)
==63426==    by 0x4910261: qcalloc (memory.c:116)
==63426==    by 0x492F7BA: prefix_new (prefix.c:1141)
==63426==    by 0x492E746: prefix_ipv6_new (prefix.c:715)
==63426==    by 0x1137E5: sharp_zebra_process_srv6_locator_chunk (sharp_zebra.c:963)
==63426==    by 0x49806E9: zclient_read (zclient.c:3992) 
==63426==    by 0x496118F: thread_call (thread.c:2005)
==63426==    by 0x48FF5E4: frr_run (libfrr.c:1198)
==63426==    by 0x110CDC: main (sharp_main.c:192)
==63426== 
==63426== 264 bytes in 1 blocks are definitely lost in loss record 27 of 38
==63426==    at 0x48455EF: calloc (vg_replace_malloc.c:1328)
==63426==    by 0x4910261: qcalloc (memory.c:116)
==63426==    by 0x11B7F1: sharp_srv6_manager_get_locator_chunk_magic (sharp_vty.c:953)
==63426==    by 0x11952C: sharp_srv6_manager_get_locator_chunk (sharp_vty_clippy.c:1619)
==63426==    by 0x48C9423: cmd_execute_command_real (command.c:997)
==63426==    by 0x48C9590: cmd_execute_command (command.c:1057)
==63426==    by 0x48C9AEA: cmd_execute (command.c:1224)
==63426==    by 0x49676F3: vty_command (vty.c:482)
==63426==    by 0x496948C: vty_execute (vty.c:1245)
==63426==    by 0x496B2B4: vtysh_read (vty.c:2144)
==63426==    by 0x496118F: thread_call (thread.c:2005)
==63426==    by 0x48FF5E4: frr_run (libfrr.c:1198)
==63426==    by 0x110CDC: main (sharp_main.c:192)

cscarpitta commented:
> I'm seeing memory leaks when I run the new tests with --valgrind-memleaks:
> 
> ==63389== 704 (344 direct, 360 indirect) bytes in 1 blocks are definitely lost in loss record 26 of 30 ==63389== at 0x48455EF: calloc (vg_replace_malloc.c:1328) ==63389== by 0x4910261: qcalloc (memory.c:116) ==63389== by 0x491208E: srv6_locator_alloc (srv6.c:132) ==63389== by 0x1EEF05: srv6_locator (zebra_srv6_vty.c:247) ==63389== by 0x48C9423: cmd_execute_command_real (command.c:997) ==63389== by 0x48C96FC: cmd_execute_command_strict (command.c:1108) ==63389== by 0x48C9B94: command_config_read_one_line (command.c:1268) ==63389== by 0x48C9D09: config_from_file (command.c:1313) ==63389== by 0x496B831: vty_read_file (vty.c:2347) ==63389== by 0x496C01F: vty_read_config (vty.c:2567) ==63389== by 0x48FED1C: frr_config_read_in (libfrr.c:984) ==63389== by 0x496118F: thread_call (thread.c:2005) ==63389== by 0x48FF5E4: frr_run (libfrr.c:1198) ==63389== by 0x1A11F3: main (main.c:475) ==63389== ==63389== 704 (344 direct, 360 indirect) bytes in 1 blocks are definitely lost in loss record 27 of 30 ==63389== at 0x48455EF: calloc (vg_replace_malloc.c:1328) ==63389== by 0x4910261: qcalloc (memory.c:116) ==63389== by 0x491208E: srv6_locator_alloc (srv6.c:132) ==63389== by 0x1EEF05: srv6_locator (zebra_srv6_vty.c:247) ==63389== by 0x48C9423: cmd_execute_command_real (command.c:997) ==63389== by 0x48C9BBC: command_config_read_one_line (command.c:1280) ==63389== by 0x48C9D09: config_from_file (command.c:1313) ==63389== by 0x496B831: vty_read_file (vty.c:2347) ==63389== by 0x496C01F: vty_read_config (vty.c:2567) ==63389== by 0x48FED1C: frr_config_read_in (libfrr.c:984) ==63389== by 0x496118F: thread_call (thread.c:2005) ==63389== by 0x48FF5E4: frr_run (libfrr.c:1198) ==63389== by 0x1A11F3: main (main.c:475) ==63389== ==63389== 704 (344 direct, 360 indirect) bytes in 1 blocks are definitely lost in loss record 28 of 30 ==63389== at 0x48455EF: calloc (vg_replace_malloc.c:1328) ==63389== by 0x4910261: qcalloc (memory.c:116) ==63389== by 0x491208E: srv6_locator_alloc (srv6.c:132) ==63389== by 0x1EEF05: srv6_locator (zebra_srv6_vty.c:247) ==63389== by 0x48C9423: cmd_execute_command_real (command.c:997) ==63389== by 0x48C9590: cmd_execute_command (command.c:1057) ==63389== by 0x48C9AEA: cmd_execute (command.c:1224) ==63389== by 0x49676F3: vty_command (vty.c:482) ==63389== by 0x496948C: vty_execute (vty.c:1245) ==63389== by 0x496B2B4: vtysh_read (vty.c:2144) ==63389== by 0x496118F: thread_call (thread.c:2005) ==63389== by 0x48FF5E4: frr_run (libfrr.c:1198) ==63389== by 0x1A11F3: main (main.c:475) ==63426== 56 bytes in 1 blocks are definitely lost in loss record 11 of 38 ==63426== at 0x48455EF: calloc (vg_replace_malloc.c:1328) ==63426== by 0x4910261: qcalloc (memory.c:116) ==63426== by 0x492F7BA: prefix_new (prefix.c:1141) ==63426== by 0x492E746: prefix_ipv6_new (prefix.c:715) ==63426== by 0x1137E5: sharp_zebra_process_srv6_locator_chunk (sharp_zebra.c:963) ==63426== by 0x49806E9: zclient_read (zclient.c:3992) ==63426== by 0x496118F: thread_call (thread.c:2005) ==63426== by 0x48FF5E4: frr_run (libfrr.c:1198) ==63426== by 0x110CDC: main (sharp_main.c:192) ==63426== ==63426== 264 bytes in 1 blocks are definitely lost in loss record 27 of 38 ==63426== at 0x48455EF: calloc (vg_replace_malloc.c:1328) ==63426== by 0x4910261: qcalloc (memory.c:116) ==63426== by 0x11B7F1: sharp_srv6_manager_get_locator_chunk_magic (sharp_vty.c:953) ==63426== by 0x11952C: sharp_srv6_manager_get_locator_chunk (sharp_vty_clippy.c:1619) ==63426== by 0x48C9423: cmd_execute_command_real (command.c:997) ==63426== by 0x48C9590: cmd_execute_command (command.c:1057) ==63426== by 0x48C9AEA: cmd_execute (command.c:1224) ==63426== by 0x49676F3: vty_command (vty.c:482) ==63426== by 0x496948C: vty_execute (vty.c:1245) ==63426== by 0x496B2B4: vtysh_read (vty.c:2144) ==63426== by 0x496118F: thread_call (thread.c:2005) ==63426== by 0x48FF5E4: frr_run (libfrr.c:1198) ==63426== by 0x110CDC: main (sharp_main.c:192)

Thanks for reporting these issues. I investigated a bit. The cause of these memory leaks is not related to this PR but to some other features already integrated in FRR, specifically the deallocation of the SRv6 locators and SRv6 nexthops.
I opened another PR to fix these memory leaks (PR https://github.com/FRRouting/frr/pull/11855).

cscarpitta commented:
> I don't see anything beyond what Donald has already commented on ... will wait for those to be resolved and look again

Thanks for the review. I already updated the PR with the implementation for all items highlighted by Donald Sharp.

Pdoijode commented on bgpd/bgp_mplsvpn.c:
```diff
@@ -1257,29 +1515,135 @@ void vpn_leak_from_vrf_update(struct bgp *to_bgp,	     /* to */
+			transposition_offset =
+				block_bits_length + node_bits_length;
+			if (chunk->function_bits_length)
+				transposition_length =
```
If the user/operator configures function length = 32 bits, then "transposition_length" field here will be set to 32, which is wrong. This is because, the number of bits that can be transposed from the function part of SID into label field (label field is 24 bits) of NLRI is 20bits only. (20 bits in case of VPNv4/v6 and the remaining 4 bits in label is used for BOS bit and EXP bits). So if the function_bits_length is > 20 bits, then set the transposition_length to 20 bits and add the rest of the bits i.e, function_bits_length - 20 to transposition offset.

From section 3.2.1 of RFC 9252:
  "As an example, consider that the sum of the Locator Block and the
   Locator Node parts is 64.  For an SRv6 SID where the entire Function
   part of size 16 bits is transposed, the transposition offset is set
   to 64 and the transposition length is set to 16.  **While for an SRv6
   SID for which the FL is 24 bits and only the lower order 20 bits are
   transposed (e.g., due to the limit of the MPLS Label field size), the
   transposition offset is set to 68 and the transposition length is set
   to 20."**

Pdoijode replied:
In addition, BGP needs to validate the that the transpose length is not greater than the label field when processing  type 5 prefix-SID attribute (BGP_PREFIX_SID_SRV6_L3_SERVICE). In case the transpose_length is > 20 bits (in case of VPN) then BGP needs to mark this path as ineligible for bestpath calculation. May be this can be done  separately in the future, but just putting it out there. 

According to "error handling" section of RFC 9252:
"The SRv6 SID value in the SRv6 SID Information Sub-TLV is invalid
   when the SID Structure Sub-Sub-TLV transposition length is greater
   than the number of bits of the label field or if any of the
   conditions for the fields of the Sub-Sub-TLV, as specified in
   [Section 3.2.1](https://datatracker.ietf.org/doc/html/rfc9252#section-3.2.1), is not met.  The transposition offset and length MUST
   be 0 when the Sub-Sub-TLV is advertised along with routes where the
   Transposition Scheme is not applicable (e.g., for global IPv6 service
   [[RFC2545](https://datatracker.ietf.org/doc/html/rfc2545)] where there is no label field).  The path having any such
   Prefix-SID attribute without any valid SRv6 SID information MUST be
   considered ineligible during the selection of the best path for the
   corresponding prefix."

cscarpitta replied:
> If the user/operator configures function length = 32 bits, then "transposition_length" field here will be set to 32, which is wrong. This is because, the number of bits that can be transposed from the function part of SID into label field (label field is 24 bits) of NLRI is upto 20bits only. (upto 20 bits in case of VPNv4/v6 since the remaining 4 bits in label is used for BOS bit and EXP bits). So if the function_bits_length is > 20 bits, then set the transposition_length to 20 bits and add the rest of the bits i.e, function_bits_length - 20 to transposition offset.
> 
> From section 3.2.1 of RFC 9252: "As an example, consider that the sum of the Locator Block and the Locator Node parts is 64. For an SRv6 SID where the entire Function part of size 16 bits is transposed, the transposition offset is set to 64 and the transposition length is set to 16. **While for an SRv6 SID for which the FL is 24 bits and only the lower order 20 bits are transposed (e.g., due to the limit of the MPLS Label field size), the transposition offset is set to 68 and the transposition length is set to 20."**

@Pdoijode Thanks for the suggestion. In my opinion, we should implement this feature/fix in a separate PR.

Note that in the current implementation of bgpd, this should not be a problem because the Function part of an SRv6 SID is limited to 8 bits (and it never exceeds 20 bits).

In bgpd, you can allocate SIDs in two ways:
* manually by providing a value in the range 1-255 that will be used as SRv6 SID Function (i.e., CLI command `sid vpn export (1-255)`)
* automatically, leaving the choice of the SID Function to bgpd (i.e., CLI command `sid vpn export auto`).

In the former case, the value that you provide cannot exceed 8 bits.
In the second case, bgpd chooses the value to use for the SRv6 SID Function in the range 1-255, and again this value never exceeds 8 bits.

So, in the current implementation, the SRv6 SID Function is limited to 8 bits, which is too limiting. We definitely need to remove this limitation.

I am already working to support SID functions larger than 8 bits, but this requires several changes to bgpd and to the bgpd CLI. I think adding these changes to this PR would be confusing and would make the review more difficult.

For this reason, I prefer to implement these changes in a separate PR. I will submit the new PR after this PR is merged. In the new PR I will also take into account that the MPLS label field is 20 bits long and will add code to handle the case you highlighted.

> In addition, BGP needs to validate the that the transpose length is not greater than the label field when processing type 5 prefix-SID attribute (BGP_PREFIX_SID_SRV6_L3_SERVICE). In case the transpose_length is > 20 bits (in case of VPN) then BGP needs to mark this path as ineligible for bestpath calculation. May be this can be done separately in the future, but just putting it out there.
> 
> According to "error handling" section of RFC 9252: "The SRv6 SID value in the SRv6 SID Information Sub-TLV is invalid when the SID Structure Sub-Sub-TLV transposition length is greater than the number of bits of the label field or if any of the conditions for the fields of the Sub-Sub-TLV, as specified in [Section 3.2.1](https://datatracker.ietf.org/doc/html/rfc9252#section-3.2.1), is not met. The transposition offset and length MUST be 0 when the Sub-Sub-TLV is advertised along with routes where the Transposition Scheme is not applicable (e.g., for global IPv6 service [[RFC2545](https://datatracker.ietf.org/doc/html/rfc2545)] where there is no label field). The path having any such Prefix-SID attribute without any valid SRv6 SID information MUST be considered ineligible during the selection of the best path for the corresponding prefix."

Yeah, that's an interesting feature. Again, in my opinion, this should be done in a separate PR.

cscarpitta replied:
@riw777 We can add a check to this PR to return fail when the function length is greater than 20 bits. Then, I will open a new PR to remove this limitation and add support for the larger function length.

cscarpitta commented:
Rebased to resolve merge conflicts.

ton31337 commented:
Can you rebase again?

cscarpitta commented:
> Can you rebase again?

@ton31337 I rebased on master. Thanks.

ton31337 commented:
Let's wait until October 4, when the freeze ends.

cscarpitta commented:
Hi @ton31337 @riw777 could we resume the work on this PR?

ton31337 reviewed (commented):
Just a trivial review check from my side, other than that, I don't have any really good insights here.

cscarpitta commented:
ci:rerun failed test_ospfv3_single_area

cscarpitta commented on bgpd/bgp_vty.c:
```diff
@@ -9666,17 +9785,23 @@ DEFPY (show_bgp_srv6,
 			tovpn4_sid ? buf_tovpn4_sid : "none");
 		vty_out(vty, "  vpn_policy[AFI_IP6].tovpn_sid: %s\n",
 			tovpn6_sid ? buf_tovpn6_sid : "none");
+		vty_out(vty, "  per-vrf tovpn_sid: %s\n",
```
Yes. Changed to use `%pI6`. Thanks for the suggestion.

https://github.com/FRRouting/frr/blob/f83fa796a2c0cf809337777304e5cb980536f2dd/bgpd/bgp_vty.c#L9790-L9795

I will open a separate PR to convert the other `vty_out`.

cscarpitta replied:
I opened a new PR to convert the other  `vty_out`: https://github.com/FRRouting/frr/pull/12070.

cscarpitta commented on bgpd/bgp_vty.c:
```diff
@@ -334,6 +334,11 @@ static int bgp_srv6_locator_unset(struct bgp *bgp)
+
+		/* refresh per-vrf tovpn_sid */
+		tovpn_sid = bgp_vrf->tovpn_sid;
+		if (tovpn_sid)
```
Fixed.

https://github.com/FRRouting/frr/blob/f83fa796a2c0cf809337777304e5cb980536f2dd/bgpd/bgp_vty.c#L337-L339

cscarpitta commented on bgpd/bgp_mplsvpn.c:
```diff
@@ -635,6 +756,143 @@ void ensure_vrf_tovpn_sid(struct bgp *bgp_vpn, struct bgp *bgp_vrf, afi_t afi)
+	if (tovpn_sid_index != 0 || tovpn_sid_auto)
+		return;
+
+	if (bgp_vrf->tovpn_sid_locator)
```
Fixed.

https://github.com/FRRouting/frr/blob/f83fa796a2c0cf809337777304e5cb980536f2dd/bgpd/bgp_mplsvpn.c#L877-L882

cscarpitta commented on bgpd/bgp_mplsvpn.c:
```diff
@@ -635,6 +756,143 @@ void ensure_vrf_tovpn_sid(struct bgp *bgp_vpn, struct bgp *bgp_vrf, afi_t afi)
+	if (tovpn_sid_index != 0 || tovpn_sid_auto)
+		return;
+
+	if (bgp_vrf->vpn_policy[afi].tovpn_sid_locator)
```
Fixed.

https://github.com/FRRouting/frr/blob/f83fa796a2c0cf809337777304e5cb980536f2dd/bgpd/bgp_mplsvpn.c#L850-L855

cscarpitta commented:
I made the following changes:
* fixed the issues reported by @ton31337 in review https://github.com/FRRouting/frr/pull/11673#pullrequestreview-1132388990
* added a check to prevent users from creating locators with function bits length greater than 20 bits (which is currently handled incorrectly in bgpd), as suggested by @riw777 in this comment: https://github.com/FRRouting/frr/pull/11673#discussion_r963738517

Here is the diff:
https://github.com/FRRouting/frr/compare/4ed71a58f8f6ce64578e1b178c1833c65936366b..f83fa796a2c0cf809337777304e5cb980536f2dd

Thanks again for the review.

cscarpitta commented:
Rebased to resolve merge conflicts.

cscarpitta commented:
Rebased to resolve merge conflicts.

cscarpitta commented:
Rebased to resolve merge conflicts.

cscarpitta commented:
Rebased to resolve merge conflicts.
